### PR TITLE
Make DetectChanges start tracking new entities (Issue #1213)

### PR DESCRIPTION
**Pull request by ajcvickers**
The only significant new code in the ChangeDetector are the TrackAddedEntities methods. The other changes are async overhead.

Notes:
- DetectChanges will only bring in the single entity added/set, not an entire graph. In the future the plan is that if the added entity is an aggregate root, then the entire aggregate will be brought in.
- Async Add is used when when DetectChangesAsync is called, but not when DetectChanges is called automatically or a notification entity triggers the change.
- Entities are always assumed to be new and put in the Added state.

More functional testing to come in a later change.

### CODE REVIEW

File: src/EntityFramework.Core/ChangeTracking/ChangeDetector.cs [hunk 1]
```diff
@@ -179,29 +178,66 @@ private void DetectPropertyChanges(StateEntry entry)
...
+            }
+        }
 
+        private Task DetectRelationshipChangesAsync(StateEntry entry, CancellationToken cancellationToken = default(CancellationToken))
+        {
             var snapshot = entry.TryGetSidecar(Sidecar.WellKnownNames.RelationshipsSnapshot);
             if (snapshot == null)
             {
-                return;
+                return Task.FromResult(0);
```
Comment (AndriySvyryd) [inline src/EntityFramework.Core/ChangeTracking/ChangeDetector.cs]: Use false

Comment (ajcvickers) [inline src/EntityFramework.Core/ChangeTracking/ChangeDetector.cs]: Oops. Sorry. You won't have to tell me again. Promise.

**Comment by AndriySvyryd** [comment]
:shipit:

**Comment by ajcvickers** [comment]
Merged.

**Pull request closed without merging**